### PR TITLE
Creating interfaces for remote logging

### PR DESCRIPTION
**Pull request by jedi22**
First draft of the enrollment plugin and part of the config plugin.

Please comment on potential structure and functionality.

They way it's designed to work is this:
Both the config and logger plugins will call the enroll getKey function. getKey
is in charge of maintaining the state of having a key and doing the enrollment.

If enroll has a key, then it will instantly return doing no HTTP requests, if it
doesn't, then it will try a few times to get a key, if it's not successful, it
fails and returns an empty string, if it succeeds, then it will return the
client enrollment key for the requester to use.

Now, if another plugin uses this enrollment key and it's request is rejected, then it can make a request with enroll=1, which will force the enrollment plugin to get a new key, even if it already has one.

### CODE REVIEW

File: include/osquery/enrollment.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+};
+
+/**
+ * @brief Enrollment plugin registry.
+ *
+ * This creates an osquery registry for "emrollment" which may implement
+ * EnrollmentPlugin. Only strings are logged in practice, and EnrollmentPlugin
+ * provides a helper member for transforming PluginRequests to strings.
+ */
+CREATE_REGISTRY(EnrollmentPlugin, "remote");
```
Comment (marpaia) [inline include/osquery/enrollment.h]: Instead of "remote", "enrollment"?

File: osquery/config/plugins/remote.cpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+#include <osquery/logger.h>
+#include <osquery/filesystem.h>
+#include <osquery/registry.h>
+
+namespace osquery {
+
+CLI_FLAG(string,
+         config_path,
+         "/var/osquery/osquery.conf",
+         "(filesystem) config plugin path to JSON config file");
```
Comment (marpaia) [inline osquery/config/plugins/remote.cpp]: Why is this here?

File: osquery/config/plugins/remote.cpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+class RemoteConfigPlugin : public ConfigPlugin {
```
Comment (marpaia) [inline osquery/config/plugins/remote.cpp]: Don't call this "Remote", that's far too general. `HTTPConfigPlugin`?

**Comment by marpaia** [comment]
## enrollment

Move `osquery/remote/plugins` to `osquery/enrollment/plugins`. Mirror the directory structure of `osquery/config` and/or `osquery/logger`. Assuming the plugin type is enrollment:

- Header at `include/osquery/enrollment.h`
- Implementation at `osquery/enrollment/enrollment.cpp`
  - This directory will have to be created
- Tests at `osquery/enrollment/enrollment_tests.cpp`
- Plugins at `osquery/enrollment/plugins/

## config plugin

Rename `osquery/config/plugins/remote.cpp` to `osquery/config/plugins/http.cpp` (or whatever you decide to call that config plugin, as per the comment above, don't use `Remote` as the name of a plugin.

File: osquery/remote/plugins/remote_enrollment.cpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+FLAG(string,
+     enrollment_endpoint,
+     "Not Specified",
+     "The endpoint for server side client enrollment");
+
+FLAG(string,
+     enrollment_port,
+     "Not Specified",
+     "The endpoint's port for server side client enrollment");
```
Comment (marpaia) [inline osquery/remote/plugins/remote_enrollment.cpp]: instead of host and port, you should just make a uri flag. cpp-netlib will handle parsing them out when you create an osquery::Request object.

File: osquery/remote/plugins/remote_enrollment.cpp [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  Status enroll();
+  std::string client_key_ = "";
+};
+
+REGISTER(RemoteEnrollmentPlugin, "remote", "enrollment");
+
+Status RemoteEnrollmentPlugin::enroll() {
+  std::string location =
+      FLAGS_enrollment_endpoint + ":" + FLAGS_enrollment_port;
+  auto r = Request<HTTPTransport, JSONSerializer>(location);
```
Comment (marpaia) [inline osquery/remote/plugins/remote_enrollment.cpp]: nix this mangling here and just have a "uri" flag.

File: osquery/remote/plugins/remote_enrollment.cpp [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+     "Not Specified",
+     "The secret to be provided to the endpoint to faciliate enrollment");
+
+class RemoteEnrollmentPlugin : public EnrollmentPlugin {
+ public:
+  std::string getKey(const PluginRequest& request);
+
+ private:
+  Status enroll();
+  std::string client_key_ = "";
```
Comment (marpaia) [inline osquery/remote/plugins/remote_enrollment.cpp]: just `std::string client_key_`

File: osquery/remote/plugins/remote_enrollment.cpp (hunk 5 above)
Comment (marpaia) [inline osquery/remote/plugins/remote_enrollment.cpp]: please include tests here to act as documentation for what you expect the server to return. see https://github.com/facebook/osquery/blob/master/osquery/remote/transports/http_tests.cpp for an example on how to create a dummy http server for use with tests.

File: osquery/remote/plugins/remote_enrollment.cpp [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+std::string RemoteEnrollmentPlugin::getKey(const PluginRequest& request) {
+  if (client_key_.length() == 0 || request.at("enroll") == "1") {
+    Status stat;
+
+    for (int i = 1; i <= MAX_TRIES; i++) {
+      stat = enroll();
+
+      if (stat.ok()) {
+        break;
+      }
```
Comment (marpaia) [inline osquery/remote/plugins/remote_enrollment.cpp]: Maybe an `else` here with a `sleep(i * i);`? Pseudo exponential backoff?

File: osquery/remote/plugins/remote_enrollment.cpp [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    }
+  }
+  return client_key_;
+}
+
+Status EnrollmentPlugin::call(const PluginRequest& request,
+                              PluginResponse& response) {
+  if (request.count("enroll") > 0) {
+    std::string key = this->getKey(request);
```
Comment (marpaia) [inline osquery/remote/plugins/remote_enrollment.cpp]: You can break out your `if`s here so that they're not nested:

```cpp
if (request.count("enroll") == 0) {
  return Status(1, "Unsupported call to enrollment plugin");
}

std::string key = this->getKey(request);
if (key == "") {
  return Status(1, "Could not enroll");
}

response.push_back({{"key", key}});
return Status(0, "OK");
```

File: osquery/remote/plugins/remote_enrollment.cpp [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    return stat;
+  }
+  // The call was ok, so store the enrolled key
+  boost::property_tree::ptree recv;
+  stat = r.getResponse(recv);
+  if (!stat.ok()) {
+    return stat;
+  }
+  if (recv.count("enrollment_key") > 0) {
+    for (const auto& node : recv.get_child("enrollment_key")) {
```
Comment (marpaia) [inline osquery/remote/plugins/remote_enrollment.cpp]: Why the loop? I'm assuming you only expect there to be one entry? Why not a zero-index and check that length == 1 in the if?

File: include/osquery/enrollment.h [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+ * Enrollment is called before any other function can be called. The purpose
+ * of enrollment is get a key from the server that uniquely identifies this host
+
+ * If the host if ever compromised and the key is no longer secret, the server
+ * may revoke the key. At which point subsequent calls to non-enrollment
+ * endpoints will fail. When this happens, enrollment should be called again
+ * to receive a new key for continued use.
+ */
+class EnrollmentPlugin : public Plugin {
```
Comment (marpaia) [inline include/osquery/enrollment.h]: since this will be subclassed, add a virtual destructor

**Comment by jedi22** [comment]
ok to test

File: osquery/enrollment/plugins/remote_enrollment.cpp [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+  Status enroll();
+  std::string client_key_;
+};
+
+REGISTER(RemoteEnrollmentPlugin, "enrollment", "enrollment");
+
+Status RemoteEnrollmentPlugin::enroll() {
+  std::string location =
+      FLAGS_enrollment_uri;
+  auto r = Request<HTTPTransport, JSONSerializer>(location);
```
Comment (marpaia) [inline osquery/enrollment/plugins/remote_enrollment.cpp]: just use `FLAGS_enrollment_uri` here instead of copying it to `location`

**Comment by jedi22** [comment]
ok to test

**Comment by theopolis** [comment]
@marpaia what are your thoughts around having a `osquery/enrollment` vs. `osquery/remote/enrollment` structure? Wont enrollment plugins always apply to remote services?

**Comment by marpaia** [comment]
@theopolis that's a great idea

**Comment by jedi22** [comment]
I moved that folder; I concur, it's a good idea.

File: CMake/CMakeLibs.cmake [hunk 12]
```diff
@@ -19,7 +19,7 @@ macro(ADD_OSQUERY_TEST IS_CORE TEST_NAME SOURCE)
     TARGET_OSQUERY_LINK_WHOLE(${TEST_NAME} libosquery)
     set(TEST_LINK_ADDITIONAL ${ARGN})
     if(NOT ${IS_CORE})
-      target_link_libraries(${TEST_NAME} libosquery_additional)
+      TARGET_OSQUERY_LINK_WHOLE(${TEST_NAME} libosquery_additional)
```
Comment (theopolis) [inline CMake/CMakeLibs.cmake]: This was a platform/testing build bug. Thanks for including the fix.

File: osquery/remote/CMakeLists.txt [hunk 13]
```diff
@@ -6,6 +6,11 @@ ADD_OSQUERY_LIBRARY(TRUE osquery_requests
   serializers/json.cpp
 )
 
+ADD_OSQUERY_LIBRARY(FALSE osquery_enrollment
+  enrollment/plugins/remote_enrollment.cpp
+)
+
 ADD_OSQUERY_TEST(TRUE requests_tests requests_tests.cpp)
 ADD_OSQUERY_TEST(TRUE json_tests serializers/json_tests.cpp)
 ADD_OSQUERY_TEST(TRUE http_tests transports/http_tests.cpp)
```
Comment (theopolis) [inline osquery/remote/CMakeLists.txt]: I'm thinking all of these tests should be build core=FALSE.

**Comment by jedi22** [comment]
ok to test

File: osquery/remote/CMakeLists.txt (hunk 13 above)
Comment (marpaia) [inline osquery/remote/CMakeLists.txt]: I don't think we should have remote networking APIs in our public API, that's not really our thing. IMO, if we want to release those kinds of APIs, we should look around something like folly, clean our stuff up and submit it there or something.

File: osquery/remote/enrollment/enrollment_tests.cpp [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+
+TEST_F(EnrollmentTests, test_enroll) {
+  // Call enroll
+  PluginRequest request = {
+      {"enroll", "1"}, // 0 enroll if needed, 1 force re-enroll
+  };
+  PluginResponse resp;
+  Status stat = Registry::call("enrollment", "enrollment", request, resp);
+  EXPECT_TRUE(stat.ok());
+  std::cout << stat.getMessage() << "\n\n\n\n\n";
```
Comment (marpaia) [inline osquery/remote/enrollment/enrollment_tests.cpp]: rm

File: osquery/remote/enrollment/enrollment_tests.cpp [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+  // Verify get key contains the string
+  EXPECT_EQ(resp.size(), 1);
+  // EXPECT_EQ(resp[0]["key"], "potatoes");
```
Comment (marpaia) [inline osquery/remote/enrollment/enrollment_tests.cpp]: uncomment?

File: osquery/remote/enrollment/enrollment_tests.cpp [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+int main(int argc, char *argv[]) {
+  testing::InitGoogleTest(&argc, argv);
+  google::InitGoogleLogging(argv[0]);
+
+  osquery::FooBarHTTPHandler handler;
+  osquery::Server::options options(handler);
+  osquery::Server server(options.address("127.0.0.1").port("8241"));
+  boost::thread t(boost::bind(&osquery::Server::run, &server));
```
Comment (marpaia) [inline osquery/remote/enrollment/enrollment_tests.cpp]: Let's talk about this tomorrow. it seems that sometimes, the thread doesn't start and launch the webserver fast enough and we're seeing connection refused errors in the tests. We may have to add a 100ms sleep or something? Or lock some static resource in a way that it would get unlocked after the http server has been started?

File: osquery/remote/enrollment/plugins/remote_enrollment.cpp [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+ private:
+  Status enroll();
+  std::string client_key_;
+};
+
+REGISTER(RemoteEnrollmentPlugin, "enrollment", "enrollment");
+
+Status RemoteEnrollmentPlugin::enroll() {
+  // auto r = Request<HTTPTransport, JSONSerializer>(FLAGS_enrollment_uri);
+  auto r = Request<HTTPTransport, JSONSerializer>("http://127.0.0.1:8241");
```
Comment (marpaia) [inline osquery/remote/enrollment/plugins/remote_enrollment.cpp]: Use `FLAGS_enrollment_uri` here

File: osquery/remote/enrollment/plugins/remote_enrollment.cpp [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+
+namespace osquery {
+
+FLAG(string,
+     enrollment_uri,
+     "Not Specified",
+     "The endpoint for server side client enrollment");
+
+FLAG(string,
+     enrollment_secret,
```
Comment (marpaia) [inline osquery/remote/enrollment/plugins/remote_enrollment.cpp]: We need to also add `app_id`, for use with OAuth applications. We can leave `enrollment_secret` named as is though.

File: osquery/remote/enrollment/enrollment_tests.cpp [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+  std::cout << stat.getMessage() << "\n\n\n\n\n";
+  // Verify get key contains the string
+  EXPECT_EQ(resp.size(), 1);
+  // EXPECT_EQ(resp[0]["key"], "potatoes");
+}
+}
+
+int main(int argc, char *argv[]) {
+  testing::InitGoogleTest(&argc, argv);
+  google::InitGoogleLogging(argv[0]);
```
Comment (marpaia) [inline osquery/remote/enrollment/enrollment_tests.cpp]: you may have to init osquery here (see core/init.cpp), but you can set `FLAGS_enrollment_uri` to be localhost here after you initialize flags.

File: osquery/remote/CMakeLists.txt (hunk 13 above)
Comment (theopolis) [inline osquery/remote/CMakeLists.txt]: Right, let's flip these to `ADD_OSQUERY_LIBRARY(FALSE...)`

File: osquery/remote/enrollment/enrollment_tests.cpp (hunk 16 above)
Comment (theopolis) [inline osquery/remote/enrollment/enrollment_tests.cpp]: You can put the next few lines into a promise ;)

File: osquery/remote/enrollment/plugins/remote_enrollment.cpp (hunk 18 above)
Comment (theopolis) [inline osquery/remote/enrollment/plugins/remote_enrollment.cpp]: Also, food for thought. Given that we don't have a good way to let plugins yield additional flags. We should be careful to design our included enrollment plugins using the language and flexibility of a very-basic-TLS-enrollment process. Adding variations that subclass a usable TLS-enrollment plugin is possible too.

**Comment by marpaia** [comment]
I like it. @theopolis?

File: osquery/remote/enrollment/plugins/http_enrollment.cpp [hunk 20]
```diff
@@ -0,0 +1,91 @@
+#include <osquery/enrollment.h>
+
+#include "osquery/remote/requests.h"
+#include "osquery/remote/transports/http.h"
+#include "osquery/remote/serializers/json.h"
+#include <boost/property_tree/json_parser.hpp>
+
+#define MAX_TRIES 5
```
Comment (theopolis) [inline osquery/remote/enrollment/plugins/http_enrollment.cpp]: This could be a gflag

File: osquery/remote/enrollment/enrollment_tests.cpp [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+  } else {
+    EXPECT_EQ(resp.size(), 1);
+  }
+}
+}
+
+int main(int argc, char *argv[]) {
+  testing::InitGoogleTest(&argc, argv);
+  google::InitGoogleLogging(argv[0]);
+  osquery::FLAGS_enrollment_uri = "http://localhost:8241";
```
Comment (theopolis) [inline osquery/remote/enrollment/enrollment_tests.cpp]: Ideally we'd not support http, but practically we should require a `--allow-remote-http` flag to make non-https requests.

File: include/osquery/enrollment.h [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+namespace osquery {
+/**
+ * @brief Superclass for enrollment plugins
+ *
+ * To interface with remote backends, first a client must be enrolled, this
+ * happens through the remote facilities and this is the first step in that
+ * process.
+
+ * Enrollment is called before any other function can be called. The purpose
+ * of enrollment is get a key from the server that uniquely identifies this host
```
Comment (theopolis) [inline include/osquery/enrollment.h]: "is to get", and + '.'

File: osquery/config/plugins/http.cpp [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+
+#include "osquery/remote/requests.h"
+#include "osquery/remote/transports/http.h"
+#include "osquery/remote/serializers/json.h"
+
+#define MAX_TRIES 5
+
+namespace osquery {
+
+DECLARE_string(app_id);
```
Comment (theopolis) [inline osquery/config/plugins/http.cpp]: Be mindful of your flag names, it's best to be slightly verbose with the name to help organize the flags and their uses. Any of `enrollment_id` or `enrollment_application_id` or `enrollment_app_id` would be more appropriate.

File: osquery/config/plugins/http.cpp [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+
+FLAG(string,
+     configuration_uri,
```
Comment (theopolis) [inline osquery/config/plugins/http.cpp]: Same, the existing `config_path` should be named `config_filesystem_path`, so maybe `config_enrollment_uri`?

File: osquery/remote/enrollment/plugins/http_enrollment.cpp [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+class RemoteEnrollmentPlugin : public EnrollmentPlugin {
+ public:
+  std::string getKey(const PluginRequest& request);
+
+ private:
+  Status enroll();
+  std::string client_key_;
+};
+
+REGISTER(RemoteEnrollmentPlugin, "enrollment", "enrollment");
```
Comment (theopolis) [inline osquery/remote/enrollment/plugins/http_enrollment.cpp]: Choose a different registry item name.

File: osquery/config/plugins/http.cpp [hunk 26]
```diff
@@ -0,0 +1,105 @@
...
+     "The endpoint for server side client enrollment");
+
+class HTTPConfigPlugin : public ConfigPlugin {
+ public:
+  Status genConfig(std::map<std::string, std::string>& config);
+};
+
+REGISTER(HTTPConfigPlugin, "config", "http");
+
+Status runEnrollment(const std::string& force = "0") {
```
Comment (theopolis) [inline osquery/config/plugins/http.cpp]: Can this force be a bool?

File: osquery/config/plugins/http.cpp [hunk 27]
```diff
@@ -0,0 +1,105 @@
...
+  return Status(0, "OK");
+}
+
+Status HTTPConfigPlugin::genConfig(std::map<std::string, std::string>& config) {
+  boost::property_tree::ptree recv;
+  for (int i = 0; i <= MAX_TRIES; i++) {
+    if (i == MAX_TRIES) {
+      return Status(1, "Could not get config");
+    }
+    if (!runEnrollment(std::to_string(i)).ok()) {
```
Comment (theopolis) [inline osquery/config/plugins/http.cpp]: `runEnrollment(i == 0).ok()`

File: osquery/config/plugins/http.cpp [hunk 28]
```diff
@@ -0,0 +1,105 @@
...
+  if (recv.count("enrollment_invalid") > 0 &&
+      recv.get<std::string>("enrollment_invalid", "") == "1") {
+    return stat;
+  }
+  return Status(0, "OK");
+}
+
+Status HTTPConfigPlugin::genConfig(std::map<std::string, std::string>& config) {
+  boost::property_tree::ptree recv;
+  for (int i = 0; i <= MAX_TRIES; i++) {
```
Comment (theopolis) [inline osquery/config/plugins/http.cpp]: This loop is a bit weird. You're unconditionally breaking and I understand why but it reads incorrectly. You should break to fail or return success on success typically. Using a continue to avoid nesting several conditionals or a long bit of code and unpleasant right-heavy block is also cool. :)

File: osquery/config/plugins/http_tests.cpp [hunk 29]
```diff
@@ -0,0 +1,108 @@
...
+                  ConfigServer::response &response) {
+    response = ConfigServer::response::stock_reply(
+        ConfigServer::response::ok,
+        std::string(
+            "{ \"schedule\": { \"launchd\": { \"query\": \"select * from "
+            "launchd;\", \"interval\": 3600 }, \"all_kexts\": { \"query\": "
+            "\"select name, version from kextstat;\", \"interval\": 600 } } "
+            "}"));
+  }
+  void log(...) {}
```
Comment (theopolis) [inline osquery/config/plugins/http_tests.cpp]: Is this needed?

File: osquery/remote/enrollment/enrollment_tests.cpp [hunk 30]
```diff
@@ -0,0 +1,73 @@
...
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+#include <thread>
+#include <iostream>
+#include <boost/network/protocol/http/server.hpp>
+#include <osquery/enrollment.h>
```
Comment (theopolis) [inline osquery/remote/enrollment/enrollment_tests.cpp]: nit, check your includes

File: osquery/remote/enrollment/plugins/http_enrollment.cpp [hunk 31]
```diff
@@ -0,0 +1,91 @@
+#include <osquery/enrollment.h>
```
Comment (theopolis) [inline osquery/remote/enrollment/plugins/http_enrollment.cpp]: Missing license/header and includes are mis-ordered.

File: osquery/remote/enrollment/plugins/http_enrollment.cpp [hunk 32]
```diff
@@ -0,0 +1,91 @@
...
+     "Not Specified",
+     "The identification number of the application");
+
+class RemoteEnrollmentPlugin : public EnrollmentPlugin {
+ public:
+  std::string getKey(const PluginRequest& request);
+
+ private:
+  Status enroll();
+  std::string client_key_;
```
Comment (theopolis) [inline osquery/remote/enrollment/plugins/http_enrollment.cpp]: nit, maybe `enrollment_key_`?

File: osquery/remote/enrollment/plugins/http_enrollment.cpp [hunk 33]
```diff
@@ -0,0 +1,91 @@
...
+  }
+  if (recv.count("enrollment_key") > 0) {
+    client_key_ = recv.get<std::string>("enrollment_key", "");
+    return Status(0, "OK");
+  } else {
+    return Status(1, "No key");
+  }
+}
+
+std::string RemoteEnrollmentPlugin::getKey(const PluginRequest& request) {
```
Comment (theopolis) [inline osquery/remote/enrollment/plugins/http_enrollment.cpp]: You can pass the a boolean here instead of the request if you're only checking for a forced re-enroll.

File: include/osquery/enrollment.h [hunk 34]
```diff
@@ -0,0 +1,48 @@
...
+class EnrollmentPlugin : public Plugin {
+ public:
+  /// The EnrollmentPlugin PluginRequest action router.
+  Status call(const PluginRequest& request, PluginResponse& response);
+
+ protected:
+  virtual std::string getKey(const PluginRequest& request) = 0;
+
+ private:
+  virtual Status enroll() = 0;
```
Comment (theopolis) [inline include/osquery/enrollment.h]: You intend this to be protected right?

Comment (jedi22) [inline include/osquery/enrollment.h]: No, it's meant to be private

File: osquery/config/plugins/http_tests.cpp (hunk 29 above)
Comment (jedi22) [inline osquery/config/plugins/http_tests.cpp]: Yeah, required by contract, otherwise
```
osquery/git/osquery/third-party/cpp-netlib/boost/network/protocol/http/server/sync_connection.hpp:57:16: error:
      no member named 'log' in 'osquery::ConfigHTTPHandler'```

File: osquery/config/plugins/http.cpp (hunk 28 above)
Comment (jedi22) [inline osquery/config/plugins/http.cpp]: Made more standard

**Comment by jedi22** [comment]
ok to test

**Comment by jedi22** [comment]
Is this good to merge now?

File: osquery/config/plugins/http_tests.cpp [hunk 35]
```diff
@@ -0,0 +1,108 @@
...
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+#include <thread>
+#include <iostream>
+#include <sstream>
+
+#include <boost/network/protocol/http/server.hpp>
+#include <boost/property_tree/ptree.hpp>
+
```
Comment (marpaia) [inline osquery/config/plugins/http_tests.cpp]: rm newline

File: osquery/config/plugins/http_tests.cpp [hunk 36]
```diff
@@ -0,0 +1,108 @@
...
+#include <gtest/gtest.h>
+
+#include <osquery/config.h>
+#include <osquery/enrollment.h>
+
```
Comment (marpaia) [inline osquery/config/plugins/http_tests.cpp]: rm newline

File: osquery/config/plugins/http.cpp [hunk 37]
```diff
@@ -0,0 +1,107 @@
...
+#include <vector>
+#include <sstream>
+
+#include <osquery/config.h>
+#include <osquery/flags.h>
+#include <osquery/logger.h>
+#include <osquery/filesystem.h>
+#include <osquery/registry.h>
+
+#include <boost/property_tree/ptree.hpp>
```
Comment (marpaia) [inline osquery/config/plugins/http.cpp]: fix these headers

**Comment by marpaia** [comment]
@osqueryer, test this, please.

**Comment by marpaia** [comment]
@osqueryer, test this, please.
